rust/origen/src/services/jtag/service.rs

```diff
@@ -2,131 +2,306 @@
+        id: usize,
+        default_ir_size: Option<usize>,
+        tclk: Option<&PinGroup>,
+        tdi: Option<&PinGroup>,
+        tdo: Option<&PinGroup>,
+        tms: Option<&PinGroup>,
+        trstn: Option<&PinGroup>,
+    ) -> Result<Service> {
+        Ok(Service {
+            id: id,
+            default_ir_size: default_ir_size,
+            tclk: {
+                if let Some(grp) = tclk {
+                    grp.to_identifier()
+                } else {
+                    ("tclk".to_string(), 0)
+                }
+            },
+            tdi: {
+                if let Some(grp) = tdi {
+                    grp.to_identifier()
+                } else {
+                    ("tdi".to_string(), 0)
+                }
+            },
+            tdo: {
+                if let Some(grp) = tdo {
+                    grp.to_identifier()
+                } else {
+                    ("tdo".to_string(), 0)
+                }
+            },
+            tms: {
+                if let Some(grp) = tms {
+                    grp.to_identifier()
+                } else {
+                    ("tms".to_string(), 0)
+                }
+            },
+            trstn: {
+                if let Some(grp) = trstn {
+                    grp.to_identifier()
+                } else {
+                    ("trstm".to_string(), 0)
+                }
+            },
+        })
+    }
+
+    pub fn reset(&self, dut: &Dut) -> Result<()> {
+        let tms = PinCollection::from_group(&dut, &self.tms.0, self.tms.1)?;
+        self._reset(&tms)
     }
 
-    pub fn write_ir(&mut self, value: Value) -> Result<()> {
-        let trans = match value {
-            Value::Bits(bits, size) => node!(
-                JTAGWriteIR,
-                match size {
-                    None => bits.len() as u32,
-                    Some(x) => x,
-                },
-                bits.data()?,
-                Some(bits.overlay_enables()),
-                bits.get_overlay()?
-            ),
-            Value::Data(value, size) => node!(JTAGWriteIR, size, value, None, None),
-        };
-        let tid = TEST.push_and_open(trans);
-
-        // This is temporary to create some AST content
-        cycles(5);
-
-        TEST.close(tid)
-            .expect("Closed JTAG write IR trans properly");
+    fn _reset(&self, tms: &PinCollection) -> Result<()> {
+        let n_id = TEST.push_and_open(node!(JTAGReset, self.id, None));
+        self.comment("Resetting JTAG Interface");
+        tms.drive_high().repeat(6);
+        TEST.close(n_id)?;
+        Ok(())
+    }
+
+    // pub fn idle(&self) -> Result<()> {
+    //     // ...
+    // }
+
+    // pub fn move_to_state(&self, state: JTAGState) -> Result<()> {
+    //     // ...
+    // }
+
+    // fn _move_to_state(&self, state: JTAGState, tms: &PinCollection) -> Result<()> {
+    //     // ...
+    // }
+
+    pub fn write_ir(&self, dut: &Dut, transaction: &Transaction) -> Result<()> {
```

Comment:
I guess it'll be the transaction data structure that needs to hold any compare data for a read while write operation

Reply:
I don't think so. I think it'd be driver specific for how it wants to handle it. A transaction itself is on a single bus, with a single data value, enables, etc. I think it'd be up to the driver accept multiple transactions and handle them accordingly. You can do what you're asking on the Rust side by getting the nodes directly for each transaction and combining them yourself. I can add some helpers for this in the future but I think that'd be a different struct altogether (`ParallelTransaction`, or something like that). It'd get kinda messy making everything in a transaction a vector then having to ensure that vector lengths are equivalent.

Can this be done in O1? I looked at the driver and everything is taking a singular `reg_or_data`.

Reply:
I think  your proposal (let the driver accept and handle multiple 1-way transactions) is the best. This is what it looks like to read while write in the o1 driver: https://github.com/Origen-SDK/origen_jtag/blob/bc4dceac90a1a560b1ebf1da2956004a69e70dc1/pattern/rww_test.rb#L13